Feature/configurable controller hotkeys

Anime37 opened the pull request:
STILL WIP (also sort of blocked by https://github.com/cemu-project/Cemu/pull/1519)

Demo: https://www.youtube.com/watch?v=CNSJTER_i8k

Added configurable controller support to configurable hotkey system.

Will add more description to this pull request when I feel like it's ready to be at least reviewed. For now, putting on draft.

Anime37 commented:
> You might want to adjust your code to fit the coding style guidelines https://github.com/cemu-project/Cemu/blob/main/CODING_STYLE.md I'd recommend the same on your other PR too

if you mean camel case vs snake notation variables, i feel like even Exzap doesn't follow those guideless (look at CemuConfig.h vairables). is it worth really worth changing?

goeiecool9999 commented:
> i wanna hear @Exzap's opinion on this, since he breaks his own rules, maybe it's the `CODING_STYLE.md` that is deprecated here.

For that particular example while the latest commit shows it's 11 months old, that commit is not the one that introduced that variable. It merely prepends ``legacy_``. The variable itself dates back to the start of the git repository.

Anime37 commented:
> You might want to adjust your code to fit the coding style guidelines https://github.com/cemu-project/Cemu/blob/main/CODING_STYLE.md I'd recommend the same on your other PR too

> > i wanna hear @Exzap's opinion on this, since he breaks his own rules, maybe it's the `CODING_STYLE.md` that is deprecated here.
> 
> For that particular example while the latest commit shows it's 11 months old, that commit is not the one that introduced that variable. It merely prepends `legacy_`. The variable itself dates back to the start of the git repository, before there was a codified coding style. You can also catch me violating the guidelines if you go through my recent commits. But that's usually when I'm making changes around existing code that violates it as well. Here are some instances of me modifying my PR's to conform after I realised it doesn't. [#1290 c795002a](https://github.com/cemu-project/Cemu/pull/1290/commits/c795002a83a2e6dd1318f0d047ab66af633e3100) [#1190 1ce9ea74](https://github.com/cemu-project/Cemu/pull/1190/commits/1ce9ea74cceb538e28a3a4e85da94bdf436a3f7d)

aight, will make the my first PR conform and then rebase this one and conform it as well.

Exzap commented:
> i wanna hear @Exzap's opinion on this, since he breaks his own rules, maybe it's the `CODING_STYLE.md` that is deprecated here.

CODING_STYLE.md mentions this in the first line:

> This document describes the latest version of our coding-style guidelines. Since we did not use this style from the beginning, older code may not adhere to these guidelines. Nevertheless, use these rules even if the surrounding code does not match.

Imho the best approach is to just do a style pass and if you miss some things thats fine. Style nitpicking can turn into bikeshedding really quickly so personally I tend to merge PRs even if they don't follow the guidelines 100%. But we will point out if a large part of the code doesn't match the guidelines.

On a side note, I didn't really have time to check out your PRs in detail yet. But a question I have is how this interacts with keyboard controller input? It's not uncommon to for people to map some controller button to space for example. Will it just trigger both in that case?

Anime37 commented:
> On a side note, I didn't really have time to check out your PRs in detail yet. But a question I have is how this interacts with keyboard controller input? It's not uncommon for people to map some controller button to space for example. Will it just trigger both in that case?

Yes, it would trigger both. But I since there is a possibility to make add modifiers to hotkeys, this problem is solvable by the user (you could set the hotkey to be `CTRL+SHIFT+ALT+SPACE` if you really wanted to).

Anime37 commented:
rebased and fixed coding style issues on latest commit

Anime37 commented:
fixed most configuration issues.
left things to do:
- clear up magic numbers (like -1 as non configured controller hotkey)
- get yelled by reviewers about polling controller input at api level

Anime37 commented:
fixed issue where clicking keyboard hotkey input button and then clicking controller hotkey input button (or vice versa) would restore the previously clicked value into a wrong input type button.

Anime37 commented:
rebased to get keyboard hotkey input cancel/clearing support. added controller hotkey input cancel/clearing via right mouse clicks

Anime37 commented:
fixed issue, where clicking the previously configured controller hotkey input button would not register.

Anime37 commented:
fixed trigger buttons not working as hotkeys.

Anime37 commented:
fixed trigger buttons not working as hotkeys.

Anime37 commented:
Added column header labels and removed keyboard modifier input button (since for keyboard modifiers keys are individual per hotkey).

![image](https://github.com/user-attachments/assets/e336b322-73a3-4732-83cb-4f15cdcefcb1)

Anime37 commented:
just coz i thought'd it be funny, added an icon for the hotkey settings window
![image](https://github.com/user-attachments/assets/406bf436-89cb-446e-8b0a-9ae442d31b81)

Anime37 commented:
moved controller support to the first PR (https://github.com/cemu-project/Cemu/pull/1519), closing.
